File: ai-service/app/ocr/language_detector.py
import re

class PythonLanguageDetector:
    """
    Python Language Detection Service for e-Bhoomi (Phase 7).
    """

    def detect_language(self, text: str) -> dict:
        if not text or not text.strip():
            return {
                "detectedLanguage": "UNKNOWN",
                "languageConfidence": 0.0,
                "detectionSource": "UNAVAILABLE",
                "teluguCharRatio": 0.0,
                "englishCharRatio": 0.0,
                "reason": "Text is empty.",
            }

        telugu_count = len(re.findall(r'[\u0C00-\u0C7F]', text))
        english_count = len(re.findall(r'[A-Za-z]', text))
        total_alpha = telugu_count + english_count

        if total_alpha == 0:
            return {
                "detectedLanguage": "UNKNOWN",
                "languageConfidence": 0.5,
                "detectionSource": "TEXT_HEURISTIC",
                "teluguCharRatio": 0.0,
                "englishCharRatio": 0.0,
                "reason": "Text contains symbols/numbers without alphabetic characters.",
            }

        te_ratio = telugu_count / total_alpha
        en_ratio = english_count / total_alpha

        if te_ratio > 0.7:
            lang = "TELUGU"
        elif en_ratio > 0.7:
            lang = "ENGLISH"
        elif te_ratio > 0.2 and en_ratio > 0.2:
            lang = "MIXED_TE"
        elif te_ratio > 0:
            lang = "TELUGU"
        else:
            lang = "ENGLISH"

        return {
            "detectedLanguage": lang,
            "languageConfidence": 0.92,
            "detectionSource": "TEXT_HEURISTIC",
            "teluguCharRatio": round(te_ratio, 2),
            "englishCharRatio": round(en_ratio, 2),
            "reason": f"Language detected as {lang}.",
        }

File: src/lib/digitization/language/languageDetector.ts
export type DocumentLanguage = 'TELUGU' | 'ENGLISH' | 'MIXED_TE' | 'UNKNOWN';
export type LanguageDetectionSource = 'TEXT_HEURISTIC' | 'OCR_METADATA' | 'MODEL' | 'UNAVAILABLE';

export interface LanguageDetectionResult {
  detectedLanguage: DocumentLanguage;
  languageConfidence: number;
  detectionSource: LanguageDetectionSource;
  teluguCharRatio: number;
  englishCharRatio: number;
  reason: string;
}

export class LanguageDetector {
  public static detectLanguage(sampleText: string): LanguageDetectionResult {
    if (!sampleText || !sampleText.trim()) {
      return {
        detectedLanguage: 'UNKNOWN',
        languageConfidence: 0.0,
        detectionSource: 'UNAVAILABLE',
        teluguCharRatio: 0,
        englishCharRatio: 0,
        reason: 'Sample text is empty or missing.',
      };
    }

    const cleanText = sampleText.trim();
    let teluguChars = 0;
    let englishChars = 0;
    let totalAlpha = 0;

    for (let i = 0; i < cleanText.length; i++) {
      const code = cleanText.charCodeAt(i);
      // Telugu Unicode Block: U+0C00 to U+0C7F
      if (code >= 0x0c00 && code <= 0x0c7f) {
        teluguChars++;
        totalAlpha++;
      }
      // Basic Latin / English: A-Z, a-z
      else if ((code >= 65 && code <= 90) || (code >= 97 && code <= 122)) {
        englishChars++;
        totalAlpha++;
      }
    }

    if (totalAlpha === 0) {
      return {
        detectedLanguage: 'UNKNOWN',
        languageConfidence: 0.5,
        detectionSource: 'TEXT_HEURISTIC',
        teluguCharRatio: 0,
        englishCharRatio: 0,
        reason: 'Text contains numbers/symbols without alphabetic characters.',
      };
    }

    const teRatio = teluguChars / totalAlpha;
    const enRatio = englishChars / totalAlpha;

    let lang: DocumentLanguage = 'UNKNOWN';
    let conf = 0.85;

    if (teRatio > 0.7) {
      lang = 'TELUGU';
      conf = Math.min(0.98, 0.85 + teRatio * 0.15);
    } else if (enRatio > 0.7) {
      lang = 'ENGLISH';
      conf = Math.min(0.98, 0.85 + enRatio * 0.15);
    } else if (teRatio > 0.2 && enRatio > 0.2) {
      lang = 'MIXED_TE';
      conf = 0.90;
    } else if (teRatio > 0) {
      lang = 'TELUGU';
      conf = 0.70;
    } else if (enRatio > 0) {
      lang = 'ENGLISH';
      conf = 0.70;
    }

    return {
      detectedLanguage: lang,
      languageConfidence: Math.round(conf * 100) / 100,
      detectionSource: 'TEXT_HEURISTIC',
      teluguCharRatio: Math.round(teRatio * 100) / 100,
      englishCharRatio: Math.round(enRatio * 100) / 100,
      reason: `Language detected as ${lang} (Telugu ratio: ${Math.round(teRatio * 100)}%, English ratio: ${Math.round(enRatio * 100)}%).`,
    };
  }
}
